pkg/providers/nutanix/validator.go

```diff
@@ -441,6 +453,243 @@ func (v *Validator) validateAdditionalCategories(ctx context.Context, client Cli
+	machineCountMap := make(map[string]int)
+	cluster := clusterSpec.Cluster.Spec
+	if cluster.ControlPlaneConfiguration.MachineGroupRef.Kind == constants.NutanixMachineConfigKind {
+		machineCountMap[cluster.ControlPlaneConfiguration.MachineGroupRef.Name] = cluster.ControlPlaneConfiguration.Count
+	}
+
+	if cluster.ExternalEtcdConfiguration != nil &&
+		cluster.ExternalEtcdConfiguration.MachineGroupRef.Kind == constants.NutanixMachineConfigKind {
+		machineCountMap[cluster.ExternalEtcdConfiguration.MachineGroupRef.Name] = cluster.ExternalEtcdConfiguration.Count
+	}
+
+	for _, workerNodeGroupConfiguration := range cluster.WorkerNodeGroupConfigurations {
+		if workerNodeGroupConfiguration.MachineGroupRef.Kind == constants.NutanixMachineConfigKind &&
+			workerNodeGroupConfiguration.Count != nil {
+			machineCountMap[workerNodeGroupConfiguration.MachineGroupRef.Name] = *workerNodeGroupConfiguration.Count
+		}
+	}
+	return machineCountMap
+}
+
+func (v *Validator) getGPUModeMapping(hosts []*v3.HostResponse) (map[int64]string, map[string]string, error) {
+	gpuDeviceIDToMode := make(map[int64]string)
+	gpuNameToMode := make(map[string]string)
+
+	for _, host := range hosts {
+		if host.Status != nil &&
+			host.Status.Resources != nil &&
+			host.Status.Resources.GPUList != nil {
+			for _, gpu := range host.Status.Resources.GPUList {
+				if gpu.DeviceID != nil {
+					gpuDeviceIDToMode[*gpu.DeviceID] = gpu.Mode
+				}
+				if gpu.Name != "" {
+					gpuNameToMode[gpu.Name] = gpu.Mode
+				}
+			}
+		}
+	}
+
+	return gpuDeviceIDToMode, gpuNameToMode, nil
+}
+
+func (v *Validator) validateGPUModeNotMixed(hosts []*v3.HostResponse, cluster *cluster.Spec) error {
+	configs := cluster.NutanixMachineConfigs
+
+	gpuDeviceIDToMode, gpuNameToMode, err := v.getGPUModeMapping(hosts)
+	if err != nil {
+		return err
+	}
+
+	gpuMode := ""
+	getGpuModeFunc := createGetGpuModeFunc(gpuDeviceIDToMode, gpuNameToMode)
+	for _, machineConfig := range configs {
+		if machineConfig.Spec.GPUs != nil {
+			for _, gpu := range machineConfig.Spec.GPUs {
+				if gpuMode == "" {
+					gpuMode = getGpuModeFunc(gpu)
+				} else {
+					if gpuMode != getGpuModeFunc(gpu) {
+						return fmt.Errorf("all GPUs in a machine config must be of the same mode, vGPU or passthrough")
+					}
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func createGetGpuModeFunc(gpuDeviceIDToMode map[int64]string, gpuNameToMode map[string]string) func(gpu anywherev1.NutanixGPUIdentifier) string {
+	return func(gpu anywherev1.NutanixGPUIdentifier) string {
+		if gpu.Type == anywherev1.NutanixGPUIdentifierDeviceID {
+			return gpuDeviceIDToMode[*gpu.DeviceID]
+		}
+
+		return gpuNameToMode[gpu.Name]
+	}
+}
+
+func (v *Validator) validateFreeGPU(ctx context.Context, v3Client Client, cluster *cluster.Spec) error {
+	res, err := v3Client.ListAllHost(ctx)
+	if err != nil || len(res.Entities) == 0 {
+		return fmt.Errorf("No GPUs found: %v", err)
+	}
+
+	if v.isGPURequested(cluster.NutanixMachineConfigs) {
```

Comment:
We should check `v.isGPURequested(cluster.NutanixMachineConfigs)` to avoid making the api call on line 657 if GPU wasn't requested

Reply:
Fixed